### ISSUE TITLE
Question: use restic as a library

### ISSUE THREAD

**Issue by GlenDC** [restic/restic#1089]
## restic (repo) version

latest master: 04d27acd60270fc63979eb7c1400c0cf6806d0e8

## Question

I was checking through the restic codebase, and it doesn't seem to be possible without having to copy a lot of the code to my own codebase to simply use restic functions directly from Go, effectively using Restic as a Library.

For example, instead of requiring restic to be installed to use a cmd such as `restic backup --stdin --stdin-filename <NAME>` I would instead prefer to call such logic directly from Go code. However when I check in the cmd code, a lot of the logic seems to be hidden behind private functions. Making it pretty impossible, without having to copy a lot of that code.

Perhaps I'm the first one to want to use restic as a library, but are there any reasons why this is not possible today? Or is it just because you didn't consider that use-case before? If you like the idea of that, I could make some time available to make it happen though. What it would mean is that the binary command functions wouldn't do much more than parsing flags and options, to than call the public restic Go API themselves, as the real command logic would in this case be moved behind the public restic Go API.

**Comment by tamalsaha** [comment]
You might find some useful info here: https://github.com/restic/restic/issues/833

**Comment by fd0** [comment]
Hey, thanks for asking. At least for now, restic is meant to be a standalone binary, that's the use case we focus on. Exposing an API as a library (that is stable) with the same guarantees as we do for the restic CLI binary requires time and effort, which we do not have at the moment. The internal structures are changing a lot, so this would be even harder than it is for the program.

We may reconsider this later, but for now: I'm sorry, but restic is not a library, and the only interface we have is the binary.

I'm going to close this issue for now, please feel free to add further comments.

**Comment by GlenDC** [comment]
@fd0 I saw you mentioning this reason before, where guaranteeing a stable API would be extra work. But I'm not sure if I fully agree with that. I'm not expecting that all internal structures are exposed. That would indeed require extra work. The binary in itself exposes an API, right? What I'm proposing is basically to have more or less the same API as a Go API, and that Go API would than be used by the binary API. So there wouldn't really be much extra work involved in having this, I would think. Or am I missing something?

Thank you for responding so fast though, appreciated greatly.

**Comment by fd0** [comment]
Ah, maybe a bit more explanation: With "API" I meant calling the `restic` CLI binary. Machine-readable output as JSON has been added already to several functions, and we can work together to add this wherever it is missing.

Although the project itself is mostly structured as a library and a binary, there are still many things handled by the code in the cmd. Turns out, backup is not so easy, and we'll need to think about how to support this in a library when we eventually come around to considering this use case.

I'm sorry that we cannot accommodate your use case right now, but I'd rather be blunt and say it than to stall and let you wonder about the state of the API :)

**Comment by carlonluca** [comment]
I'm also developing a GUI for restic. I use the CLI passing params and parsing the output. Is there any guarantee about the format of the output? For instance, can you guarantee you won't change within the same minor version for instance?

**Comment by mholt** [comment]
@carlonluca I don't think so, but as of late, more and more commands are supporting the `--json` global flag which is useful for this (I'm _also_ writing a GUI for restic).

**Comment by carlonluca** [comment]
Yes, I read about that, thanks. I tried but couldn't get any json output yet. I'm probably doing something wrong.
In any case the question remains: can restic guarantee some kind of contract? I tried and it seemed the output changed from 0.8 to 0.9, but seemed to remain constant in 0.9.n. I could, for instance, support 0.8.n and 0.9.n, implementing whatever interface I need (be it json or natural language), but it depends on what policy restic intends to use in the future.
Of course this is not an elegant type of IPC, but I'm taking as assumption that restic won't implement anything better soon, as stated multiple times.
